Add performance tuning ops guide (#3731)
istio/istio.io: content/help/ops/perf/index.md

---
title: Performance Tuning
description: Istio performance, sizing and tuning guide.
weight: 90
force_inline_toc: true
---
## Istio Performance Tuning

Istio's goal is to have a performant out of the box configuration.
However it is difficult to create a configuration that is performant for all use cases.

The following section documents a few configuration knobs that affect performance.

> | Knob | Default | Symptom | Description |

Reviewer: What about 'Performance Impact' instead of 'Suspicious Behavior' ?

| --- | --- | --- | --- |
| [`Sidecar resource`](/docs/reference/config/networking/v1alpha3/sidecar) | Disabled | High memory usage by sidecars and high CPU by Pilot | Enable namespace isolation to get a significant reduction in sidecar memory usage and pilot CPU usage. |

> | `global.proxy.concurrency` | `2` | Sidecar resource utilization and latency.| This values should be at least equal to the number of CPU core allocated to the proxy. |

Reviewer 2: ```suggestion
| `global.proxy.concurrency` | `2` | An incorrect value can cause high sidecar resource use and latency. | Set this value to equal to the number of CPU cores allocated to the proxy. |
```

> | `pilot.keepaliveMaxServerConnectionAge` | `30m` | Uneven load distribution across Pilot replicas. | Reduce this number if you want quicker balancing of load at the cost of increased system churn. |

Reviewer 2: The "symptom" column ca
```suggestion
| `pilot.keepaliveMaxServerConnectionAge` | `30m` | A long time can cause uneven load distribution across Pilot replicas. | Reduce this number if you want quicker balancing of load at the cost of increased system churn. |
```

Reviewer: The uneven distribution may happen anyway, but if a shorter keepalive age is used, this situation will be fixed faster. What about 'Uneven load distribution across Pilot replicas for a longer period of time'?

> | `pilot.keepaliveMaxServerConnectionAge` | `30m` | Uneven load distribution across Pilot replicas. | Reduce this number if you want quicker balancing of load at the cost of increased system churn. |

Reviewer 3: What does "increased system churn" mean here? Is that defined somewhere?

| `pilot.traceSampling` | `1.0` |  Latency and throughput impact due to tracing. | This can be reduced further from the default. If set to 0.0%, use `x-client-trace-id` header to trace specific requests. |

> | `global.disablePolicyChecks` | `true` | Tail latency impact due to policy. | Enable policy checks *only* if you are using rate limiting or another istio-policy feature.|

Reviewer 3: Is there a page that can be linked to that defines what istio-policy features exist?

Reviewer 2: There are two possible links in the conceptual and reference docs: 
- https://preliminary.istio.io/docs/concepts/policies-and-telemetry/
- https://preliminary.istio.io/docs/reference/config/istio.authentication.v1alpha1/
That said, I recommend pointing to the rate limiting task since that is the feature mentioned:
https://preliminary.istio.io/docs/tasks/policy-enforcement/rate-limiting/

---
title: Performance Tuning
description: Istio performance, sizing and tuning guide.
weight: 90
force_inline_toc: true
---
## Istio Performance Tuning

Istio's goal is to offer a performance-optimized configuration out of the box.
However, each use case has different performance criteria. Thus, creating performance-optimized configurations for all use cases is difficult.

The following table shows a few configuration options that affect performance:

| Key | Default Value | Performance Impact | Recommendation |
| --- | --- | --- | --- |
| `global.proxy.concurrency` | `2` | An incorrect value can cause high sidecar resource use and latency. | Set this value to equal to the number of CPU cores allocated to the proxy. |

> | `pilot.keepaliveMaxServerConnectionAge` | `30m` | Uneven load distribution across Pilot replicas for a longer period of time. | Reduce this number if you want quicker balancing of load at the cost of increased connect/disconnect activity. |

Reviewer 4: I think it is odd for us to have `reduce this number` under recommendation. why would a user wants quicker balancing of load?

Reviewer: @linsun Suppose a mesh with many proxies and 1 Pilot instance initially. As the HPA kicks off and new Pilot instances are added, the proxies will continue to be connected to the original one until the connection age expires, maybe overloading a single Pilot. By tuning the connection age, this transition can be quicker, minimizing this situation.

Another important configuration option that should be is the considered for tuning is the [`Sidecar resource`](/docs/reference/config/networking/v1alpha3/sidecar), which is disabled by default. In a large mesh, there will be a high memory usage by sidecars and high CPU by Pilot, if the sidecar scope includes the whole mesh. By enabling namespace isolation there should be a significant reduction in sidecar memory usage and pilot CPU usage.

### Istio default profile vs. functionality

In order to get a good out of the box performance, Istio 1.1 disabled several features by default.

| Feature | Default | Description |
| --- | --- | --- |
| `global.proxy.accessLogFile` | `""` |  Proxy access logging is disabled because it contributes to high CPU and I/O. Set this to `/dev/stdout` to enable access logging.|
| `mixer.adapters.stdio.enabled` | `false` |  Mixer access logging is disabled because it is a debug feature and it contributes to high CPU and I/O. Set this to `true` to enable access logging.|

> | `global.disablePolicyChecks` | `true` | Enable policy checks *only* if you are using rate limiting or another istio-policy feature.|

Reviewer 4: this flag is mentioned earlier, do we need to duplicate it again here?

Reviewer: It is good to re-enforce that this feature is disabled by default, but agree that description should be different than the previous table.

### Namespace isolation

Namespace isolation using the sidecar resource is a major performance improvement in Istio 1.1
over the previous release of Istio. When namespace isolation is turned on-

> 

Reviewer 4: do we have docs on how to enable it?  or at least link to an existing doc?

Reviewer: There is a link already in the previous section, but we can repeat it here.

| Pilot CPU      | 1 vCPU | 5 vCPU | 1 vCPU | 8 vCPU |
| Pilot Memory   | 1.4 GB | 4 GB | 1.4 GB | 8 GB |
| Proxy Memory (per instance) | 50 MB  | 170 MB | 50 MB  | 225 MB |

### Istio Sizing and Capacity Planning Guidance

Based on all the performance data available, one can have an educated sizing and capacity planning decisions for Istio. The capacity planning effort depends on the kind of application, transaction volume and what features of Istio are enabled among many other factors. It can be looked at from how much total throughput the application has to support at what latency and can also be viewed as how many requests pass through the mesh. The Istio Performance Grafana dashboard is updated to provide this data for any workload using Istio.

The following table summarizes the resources requirements for Istio main components for the default configuration:

| Istio Component | CPU | Memory | Condition |
| --- | --- | --- | --- |
| Proxy | 0.6 vCPU | 50 MB `*` | CPU: 1000 mesh requests/sec; Mem: 1000 services, 2000 sidecars |
| Telemetry | 0.6 vCPU | 200 MB | 1000 mesh requests/sec |
| Pilot | 1 vCPU | 1500 MB | 1000 services, 2000 sidecars |

> 

Reviewer 4: is this with ns isolation on or off?  by default, it is off.

Reviewer: NS isolation is on. Will make a remark.

### Istio default profile vs. functionality

In order to get a good performance out of the box, Istio 1.1 disabled several features by default:

| Feature | Default | Description |
| --- | --- | --- |
| `global.proxy.accessLogFile` | `""` |  Proxy access logging is disabled because it contributes to high CPU and I/O. Set this to `/dev/stdout` to enable access logging.|
| `mixer.adapters.stdio.enabled` | `false` |  Mixer access logging is disabled because it is a debug feature and it contributes to high CPU and I/O. Set this to `true` to enable access logging.|
| `global.disablePolicyChecks` | `true` | Policy checks are disabled by default to avoid unnecessary latency in the data path. Set it to `true` in case of traffic management policies are added. |

> ### Namespace isolation

Reviewer 2: Please make all headings second level headings:
```suggestion
## Namespace isolation
```

- The default connectivity is namespace wide. Previously it was mesh wide.
- Pilot configures envoy with information that is relevant to routing within the namespace.
- Proxy receives less configuration information, therefore the proxy memory usage is lower.

The following tables documents the reduction in resource utilization due to namespace isolation for 2 different setups.

Configuration of each setup:

| Parameter | Setup 1 | Setup 2 |
| --- | --- | --- |
| Namespaces (N)| 40 | 50 |
| Sidecars (N x 40) | 1600 | 2000 |
| Services (N x 20) | 800 | 1000 |
| Virtual Services (N x 2) | 80 | 100|
| Gateways (N) | 40 | 50 |
| Configuration change rate (N/4 per minute) | 10/min | 12/min |

, and a comparison of resource utilization for each, with and without namespace isolation:

| Resource | Setup 1 || Setup 2 ||
| --- | --- | --- | --- | --- |
| Namespace Isolation | Enabled |  Disabled | Enabled | Disabled |
| Pilot replicas | 1  | 2 | 1 | 3 |
| Pilot CPU      | 1 vCPU | 5 vCPU | 1 vCPU | 8 vCPU |
| Pilot Memory   | 1.5 GB | 4 GB | 1.5 GB | 8 GB |
| Proxy Memory (per instance) | 50 MB  | 170 MB | 50 MB  | 225 MB |

> ### Istio Sizing and Capacity Planning Guidance

Reviewer 2: Capitalization issue and avoid redundancy. Readers already know they are in the Istio docs so you don't need to include it in the headings.

```suggestion
### Size and capacity planning
```

- The default connectivity is namespace-wide instead of mesh-wide.
- Pilot configures Envoy with only the relevant information to route traffic within the namespace.
- The Envoy proxy receives less configuration information lowering the proxy's memory usage.

The following table document shows the configuration parameters of two different meshes: Setup 1 and Setup 2.

| Parameter | Setup 1 | Setup 2 |
| --- | --- | --- |
| Namespaces (N)| 40 | 50 |
| Sidecars (N x 40) | 1600 | 2000 |
| Services (N x 20) | 800 | 1000 |
| Virtual Services (N x 2) | 80 | 100|
| Gateways (N) | 40 | 50 |
| Configuration change rate (N/4 per minute) | 10/min | 12/min |

The following table shows the impact enabling namespace isolation has in the resource consumption of Setup 1 and Setup 2:

| Resource | Setup 1 || Setup 2 ||
| --- | --- | --- | --- | --- |
| Namespace Isolation | Enabled |  Disabled | Enabled | Disabled |
| Pilot replicas | 1  | 2 | 1 | 3 |
| Pilot CPU      | 1 vCPU | 5 vCPU | 1 vCPU | 8 vCPU |
| Pilot Memory   | 1.5 GB | 4 GB | 1.5 GB | 8 GB |
| Proxy Memory (per instance) | 50 MB  | 170 MB | 50 MB  | 225 MB |

## Size and capacity planning

> Using the available the performance data, you can make educated sizing and capacity planning decisions for your Istio mesh.

Reviewer 5: "Using available performance data," ...

Your capacity planning effort depends on the following factors among many others:

> * The kind of application running in the mesh.

Reviewer 5: change `*` to `-` which will be a bulleted list.

* The transaction volume of the mesh.
* The Istio features enabled. 

> You need to consider capacity from many angles: 

Reviewer 5: `Capacity should be considered from many vectors: `might be better.  Angle is an idiom - vector is not.

> * How much total throughput the application has to support?

Reviewer 5: ```
- Application throughput requirements.
- Application latency requirements.
- Mesh RPS (Requests Per Second) throughput requirements.
```

* At what latency?
* How many requests pass through the mesh? 

> We updated the Grafana dashboard for Istio performance to provide this data for any workload within your Istio mesh.

Reviewer 5: "The Grafana dashboard" - shouldn't use `we` in documentation.

The following table shows the resource requirements for Istio's main components in a typical mesh configuration:

| Istio Component | CPU | Memory | Condition |
| --- | --- | --- | --- |

> | Proxy | 0.6 vCPU | 50 MB per instance | CPU: 1000 mesh requests per second (rps); Memory: 1000 services, 2000 sidecars with namespace isolation enabled. |

Reviewer 5: This could be dangerous - I'd recommend indicating how many services were isolated by sidecar isolation.

| Telemetry | 0.6 vCPU | 200 MB | 1000 mesh requests/sec |
| Pilot | 1 vCPU | 1.5 GB | 1000 services, 2000 sidecars with namespace isolation enabled. |

> 

Reviewer 5: Need a conclusion of some sort :)
